Use Devices instead of DeviceIndexes in TensorPipe agent

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57354 Inline hooks in ivalue::Future
* #57353 Use Devices instead of DeviceIndexes in Future
* **#57294 Use Devices instead of DeviceIndexes in TensorPipe agent**
* #57345 Rename DeviceGuardImplInteface's getStreamFromPool method
* #57293 Add device() method to c10::Event
* #57292 Add pybind type caster for c10::Device

With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

lw committed:
Use Devices instead of DeviceIndexes in TensorPipe agent

With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -103,29 +104,30 @@ std::unordered_set<c10::DeviceIndex> getLocalDevices(
 
   // opts_.devices must be a superset of local devices in reverseDeviceMaps_
-  std::vector<c10::DeviceIndex> cut;
-  std::set_difference(
```
I believe `set_difference` requires its inputs to be sorted, so I don't understand how this used to work with `unordered_set` which, as its name says, isn't sorted...

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -708,11 +709,9 @@ void TensorPipeAgent::pipeRead(
+    std::vector<c10::Device>&& devices,
     std::shared_ptr<LazyStreamContext> ctx,
-    std::function<void(const tensorpipe::Error&)> fn,
-    const std::unordered_map<c10::DeviceIndex, c10::DeviceIndex>&
-        deviceMap) noexcept {
```
The `deviceMap` argument was unused, so I removed it.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -37,14 +37,14 @@ namespace channel {
 class Context;
 } // namespace channel
 
-using DeviceMap = std::unordered_map<c10::DeviceIndex, c10::DeviceIndex>;
-
```
Not sure if there was a reason to define this in the `tensorpipe` namespace, which was otherwise used just for forward-declaring TensorPipe's public API. I think it makes more sense for this to be in the `torch` namespace.

lw commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -142,50 +127,37 @@ struct TORCH_API LazyStreamContext {
 
-  std::set<c10::DeviceIndex> devices() const {
-    std::set<c10::DeviceIndex> devices;
+  std::unordered_set<c10::Device> devices() const {
```
Is this method being used anywhere? There didn't seem to be any consequences to changing the return type...

mrshenli replied:
it's used here:

https://github.com/pytorch/pytorch/blob/65968ab817db323a532f50a2f2ea131ae27dada5/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L758

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -268,7 +268,7 @@ std::shared_ptr<JitFuture> ProcessGroupAgent::send(
     Message&& message,
     const float rpcTimeoutSeconds,
-    const std::unordered_map<c10::DeviceIndex, c10::DeviceIndex>& deviceMap) {
+    const std::unordered_map<c10::Device, c10::Device>& deviceMap) {
```
this is prior to this PR. IIUC, this is /* unused */

mrshenli commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -142,50 +127,37 @@ struct TORCH_API LazyStreamContext {
+  std::unordered_map<c10::Device, c10::Stream> streams_;
 };
 
 inline std::shared_ptr<LazyStreamContext> createLazyStreamContext(
```
do we still need this helper? looks like we can directly call the `LazyStreamContext` ctor?

mrshenli reviewed (approved):
Build failures are real. But since looks like it is only caused by test code and easy to fix, stamp to unblock.  Please wait for tests.  Thanks for fixing and improving!

```
In file included from /pytorch/test/cpp/rpc/test_tensorpipe_serialization.cpp:5:0:
/pytorch/torch/csrc/distributed/rpc/tensorpipe_utils.h:51:1: note: declared here
 tensorpipeSerialize(
 ^~~~~~~~~~~~~~~~~~~
/pytorch/test/cpp/rpc/test_tensorpipe_serialization.cpp: In member function ‘virtual void TensorpipeSerialize_NoDeleterTensors_Test::TestBody()’:
/pytorch/test/cpp/rpc/test_tensorpipe_serialization.cpp:153:39: error: too few arguments to function ‘std::tuple<tensorpipe::Message, torch::distributed::rpc::TensorpipeWriteBuffers> torch::distributed::rpc::tensorpipeSerialize(torch::distributed::rpc::Message&&, std::vector<c10::Device>, const std::shared_ptr<torch::distributed::rpc::LazyStreamContext>&)’
           std::move(sendingRpcMessage));
                                       ^
In file included from /pytorch/test/cpp/rpc/test_tensorpipe_serialization.cpp:5:0:
/pytorch/torch/csrc/distributed/rpc/tensorpipe_utils.h:51:1: note: declared here
 tensorpipeSerialize(
```

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]

lw committed:
Update on "Use Devices instead of DeviceIndexes in TensorPipe agent"


With the advent of CPUs in the device maps, and to be more generic (e.g., to support AMD GPUs), and to avoid conversions when passing to Future and RRef and such, it's easier to use Devices instead of DeviceIndices. This started by just migrating the TensorPipe agent but the RPC layer is quite intertwined so I had to migrate a lot of stuff.

Differential Revision: [D28092733](https://our.internmc.facebook.com/intern/diff/D28092733/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28092733/)!

[ghstack-poisoned]
